File: Java/src/test/java/linkedlist/DoublyLinkedList/DoublyLinkedListTest.java
package linkedlist.DoublyLinkedList;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.example.linkedlist.DoublyLinkedList.DoublyLinkedList;

public class DoublyLinkedListTest {

    @Test
    public void testDoublyLinkedList() {
        DoublyLinkedList list = new DoublyLinkedList();
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);
        list.append(6);
        list.append(7);
        list.append(8);
        list.append(9);
        list.append(10);
        Assertions.assertTrue(list.size() == 10);
        Assertions.assertTrue(list.findAt(0) == 1);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1 });
    }

    @Test
    public void testDoublyLinkedListPrepend() {
        DoublyLinkedList list = new DoublyLinkedList();
        list.prepend(1);
        list.prepend(2);
        list.prepend(3);
        Assertions.assertEquals(list.size(), 3);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 3, 2, 1 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 1, 2, 3 });
    }

    @Test
    public void testDoublyLinkedListInsertAt() {
        DoublyLinkedList list = new DoublyLinkedList();
        list.insertAt(1, 0);
        list.insertAt(2, 1);
        list.insertAt(3, 2);
        Assertions.assertEquals(list.size(), 3);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 1, 2, 3 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 3, 2, 1 });

        list.append(4);
        list.insertAt(5, 3);
        Assertions.assertEquals(list.size(), 5);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 1, 2, 3, 5, 4 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 4, 5, 3, 2, 1 });

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.insertAt(6, 7);
        });
        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.insertAt(6, -1);
        });
    }

    @Test
    public void testDoublyLinkedListDeleteAt() {
        DoublyLinkedList list = new DoublyLinkedList();
        list.append(1);
        list.append(2);
        list.append(3);
        list.append(4);
        list.append(5);
        list.append(6);
        Assertions.assertEquals(list.size(), 6);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 1, 2, 3, 4, 5, 6 });
        list.deleteAt(0);
        Assertions.assertEquals(list.size(), 5);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 2, 3, 4, 5, 6 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 6, 5, 4, 3, 2 });

        list.deleteAt(3);
        Assertions.assertEquals(list.size(), 4);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 2, 3, 4, 6 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 6, 4, 3, 2 });

        list.deleteAt(3);
        Assertions.assertEquals(list.size(), 3);
        Assertions.assertArrayEquals(list.toArray(), new int[] { 2, 3, 4 });
        Assertions.assertArrayEquals(list.reverseToArray(), new int[] { 4, 3, 2 });
        Assertions.assertFalse(list.isEmpty());
        Assertions.assertTrue(list.findAt(0) == 2);

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(3);
        });

        Assertions.assertThrows(IndexOutOfBoundsException.class, () -> {
            list.deleteAt(-1);
        });
    }

    @Test
    public void testDoublyLinkedListFindAt() {
        DoublyLinkedList list = new DoublyLinkedList();
        list.append(1);
        list.append(2);
        list.append(3);
        Assertions.assertEquals(list.findAt(0), 1);
        Assertions.assertEquals(list.findAt(1), 2);
        Assertions.assertEquals(list.findAt(2), 3);
    }
}

File: go/linkedlist/DoublyLinkedList/DoublyLinkedList_test.go
package doublylinkedlist_test

import (
	"testing"

	linkedlist "example.com/go/go/linkedlist/DoublyLinkedList"

	"gotest.tools/v3/assert"
)

func TestDoublyLinkedListIsEmpty(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	assert.Equal(t, d.IsEmpty(), true)

}

func TestDoublyLinkedListAppend(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	d.Append(1)
	d.Append(2)
	d.Append(3)
	assert.Equal(t, d.Size(), 3)
	assert.DeepEqual(t, d.ToArray(), []int{1, 2, 3})
	assert.DeepEqual(t, d.ReverseToArray(), []int{3, 2, 1})
}

func TestDoublyLinkedListPrepend(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	d.Prepend(1)
	d.Prepend(2)
	d.Prepend(3)
	d.Prepend(4)
	assert.Equal(t, d.Size(), 4)
	assert.DeepEqual(t, d.ToArray(), []int{4, 3, 2, 1})
	assert.DeepEqual(t, d.ReverseToArray(), []int{1, 2, 3, 4})
}

func TestDoublyLinkedListResverse(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	d.Append(1)
	d.Append(2)
	d.Append(3)
	d.Append(4)
	assert.DeepEqual(t, d.ToArray(), []int{1, 2, 3, 4})
	d.Reverse()
	assert.DeepEqual(t, d.ToArray(), []int{4, 3, 2, 1})

}

func TestDoublyLinkedListInsertAt(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	d.Append(1)
	d.Append(2)
	d.Append(3)
	d.InsertAt(4, 1)
	assert.DeepEqual(t, d.ToArray(), []int{1, 4, 2, 3})
	assert.DeepEqual(t, d.ReverseToArray(), []int{3, 2, 4, 1})
	d.InsertAt(5, 0)
	assert.DeepEqual(t, d.ToArray(), []int{5, 1, 4, 2, 3})
	assert.DeepEqual(t, d.ReverseToArray(), []int{3, 2, 4, 1, 5})

	d.InsertAt(6, 5)
	assert.Equal(t, d.Size(), 6)
	assert.DeepEqual(t, d.ToArray(), []int{5, 1, 4, 2, 3, 6})
	assert.DeepEqual(t, d.ReverseToArray(), []int{6, 3, 2, 4, 1, 5})
}

func TestDoublyLinkedListDeleteAt(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	d.Append(1)
	d.Append(2)
	d.Append(3)
	d.Append(4)
	d.Append(5)
	d.Append(6)
	assert.DeepEqual(t, d.ToArray(), []int{1, 2, 3, 4, 5, 6})
	d.DeleteAt(0)
	assert.Equal(t, d.Size(), 5)
	assert.DeepEqual(t, d.ToArray(), []int{2, 3, 4, 5, 6})
	assert.DeepEqual(t, d.ReverseToArray(), []int{6, 5, 4, 3, 2})
	d.DeleteAt(4)

	assert.Equal(t, d.Size(), 4)
	assert.DeepEqual(t, d.ToArray(), []int{2, 3, 4, 5})
	assert.DeepEqual(t, d.ReverseToArray(), []int{5, 4, 3, 2})
	d.DeleteAt(2)
	assert.Equal(t, d.Size(), 3)
	assert.DeepEqual(t, d.ToArray(), []int{2, 3, 5})
	assert.DeepEqual(t, d.ReverseToArray(), []int{5, 3, 2})
	d.DeleteAt(2)
	assert.Equal(t, d.Size(), 2)
	assert.DeepEqual(t, d.ToArray(), []int{2, 3})
	assert.DeepEqual(t, d.ReverseToArray(), []int{3, 2})

}

func TestIndexBoundsError(t *testing.T) {
	d := linkedlist.NewDoublylinkedlist()
	assert.Assert(t, d.IsEmpty())
	d.InsertAt(1, 2)
}
